include/kernel.h

```diff
@@ -527,7 +527,25 @@ __syscall void k_wakeup(k_tid_t thread);
  * @return ID of current thread.
  *
  */
-__syscall k_tid_t k_current_get(void);
+__attribute__((const)) __syscall k_tid_t k_current_get(void);
+
+static ALWAYS_INLINE __attribute__((const)) k_tid_t z_impl_k_current_get(void)
+{
+#ifdef CONFIG_SMP
+	/* In SMP, _current is a field read from _current_cpu, which
+	 * can race with preemption before it is read.  We must lock
+	 * local interrupts when reading it.
+	 */
+	unsigned int k = arch_irq_lock();
```

Comment:
FWIW: I'd still rather see this be a bare read of _current_cpu (which already contains an assert to assure atomicity) and then fix up any offending users to either move it into an existing lock or use k_current_get().  The majority of the time, we don't need to be taking this lock.

Though maybe we should just move this into an arch_current_get() and have the architecture make the decision as to optimization.  Note that on x86 with a little work we could have %gs point to the struct _cpu (it currently points to a TSS with an extra struct after it, I think?) and read the _current field out in the same mov instruction, atomically, with no locks at all.